README.md
# segcryptor
This is a library of functions for use in encrypting passwords, generating UUIDs, generating random strings of hexadecimal data, or even generating a random alphanumeric key.  Any encryption is handled by the SHA-512 encryption algorithm.

## makeID(options)

## makeHash(options)
Generates a string of random hexadecimal characters for a given number of bytes, where `options.size` is the number of bytes. Note that there are 2 characters per byte, so the string length will be twice the `options.size` value.

## makeKey(options)

## encryptPassword(hash, salt, options)

## saltHash(hash, salt)

## encrypt(string, key, iv)

## decrypt(data, key, iv)

## hashValue(string, options)
Takes a `string` of data and applies a SHA-512 encryption hashing algorithm to it to generate a string of 128 hexadecimal characters.

## toBoolean(value)

## randomBytes(size)
Generates a buffer of random hexadecimal data, where `size` is a given integer for the number of pairs of random hexadecimal data to be generated.

## Note
Note that if you use `makeID()` the `version` option needs to be a string because `1.0` and `1` are seen as the same value when interpreted as a number, whereas the strings of `'1.0'` and `'1'` are not the same value


CHANGELOG.md
## CHANGELOG

### 1.6.2: reorganization
- reorganized most of the code
- modified `hashValue()` to no longer use nodejs-specific APIs - it should now work on frontend or backend
- modified `webHashValue()` to point to `hashValue()` to maintain backwards compatibility, as there's no longer any need for a distinction between the two functions
- added test examples for `hexToLatin()` and `latinToHex()` to `./src/tests.js`

### 1.6.1: webHashValue()
- added `webHashValue()` for applying a SHA-512 hashing algorithm to a string on the frontend - `hashValue()` uses nodejs-specific APIs and will only work on the backend
- added tests for `encrypt()` and `decrypt()`

### 1.5.7: makeHash makeKey reimplemented
- fixed an issue with the last commit where my linter decided to randomly remove `makeHash()` and `makeKey()`

### 1.5.6: module.exports restructure
- restructured how functions are exported to try to fix the previous issues and maybe clean up the code a little

### 1.5.5: another attempt
- trying once again to fix some dumb issue with functions that call `toBoolean()` not working properly in other projects

### 1.5.4: reverted 1.5.2
- reverted the changes in 1.5.2

### 1.5.3: tests and rearrangement
- moved the `toBoolean()` function to the top of the code
- added tests for `toBoolean()`

### 1.5.2: makeKey toBoolean fix
- fixed an issue with `makeKey()` where it was trying to call `toBoolean()` on `this` even though `this` wasn't applicable in that context

### 1.5.1: hexadecimal conversion functions
- added `hexToLatin()` for converting a string of hexadecimal characters to plain-text latin encoded characters
- added `latinToHex()` for converting a string of plain-text latin encoded characters to a string of hexadecimal characters

### 1.4.1: hashValue improvements
- modified `hashValue()` to use the integrated `crypto` API for handling hashing SHA encryptions
- deleted '/sha.js'
- modified `hashValue()` to accept a `type` string property on the `options` for defining which hashing algorithm to use - `'sha256'` for SHA-256 and `'sha512'` for SHA-512

### 1.3.1-1: 256-bit v0 hybrids
- added `0.1`, `1.0`, `0.3`, `3.0`, `0.4`, `4.0`, `0.5`, and `5.0` versions for `makeID()`
- added tests for the new versions of `makeID()`
- modified the `version` option in `makeID()` to require a string as opposed to using a number, since new versions may get confused due to how numbers are interpreted by javascript

### 1.2.3-1: hashValue and other fixes
- fixed issues where functions where attempting to call `hashString()` which was changed to `hashValue()`
- made it so that `makeID256()` and any future functions that aren't exported have an easier time accessing functions that are exported
- updated `eslint` to `7.19.0`
- updated `uuid` to `8.3.2`
- made a minor adjustment to the uuid-checking regular expression
- fixed an issue with uuidv5.1 generation where the first 18 characters of the first part weren't being selected and the entire string was being used instead

### v1.2.2-1: randomBytes wrapper
- added `randomBytes()`, a wrapper for `crypto.randomBytes()` as instances where you would use `encrypt()` or `decrypt()` would commonly want to use that function to generate random bytes of data

### v1.2.1-1: encrypt and decrypt
- added `encrypt()` which takes a string and encrypts it using the `'aes-256-cbc'` algorithm, given a key and IV
- added `decrypt()` which decrypts an `'aes-256-cbc'` encrypted string, given a key and IV
- added a 'test' script to 'package.json'

### 4: hashString renamed
- renamed `hashString()` to `hashValue()` and updated any references to it
- modified `hashValue()` to accept an `options` parameter instead of the `count` parameter, and the `options` parameter has a property for `count`
- improved the way `saltHash()` works

### 3: encryptPassword security
- changed the `security` property on the `options` parameter for `encryptPassword()` to `count` and made it so it is no longer exponentially applied

### 2: optimized makeKey()
- optimized the code for `makeKey()`

### 1: initial commit
- initial commit

---

## TODO

### modify various functions to use Node.js's integrated 'crypto' module for creating SHA-256 and SHA-512 hashes

### add an option to `hashValue()` for using the SHA-256 algorithm instead of only the SHA-512 algorithm

### add `buffer` and `offset` options to v3, v1.3, v3.1, v5, v1.5, and v5.1

### add an option for removing hyphens from the resulting UUID-generated strings


src/tests.js
const {
  makeID, makeHash, makeKey, encryptPassword, hashValue, saltHash, toBoolean, encrypt, decrypt, hexToLatin, latinToHex, webHashValue,
} = require(`../src/index`);

/******************************************************************************/

// test generating a UUIDv1
const exampleMakeIDv1 = makeID({
  version: `1`,
});
console.log(`exampleMakeIDv1: ${exampleMakeIDv1}\r\n`);

/******************************************************************************/

// test generating a UUIDv3
// the namespace option can be
// a UUID, like '12345678-1234-1234-1234-123456789012'
// a URL, like 'https://google.com'
// a DNS address, like 'google.com'
const exampleMakeIDv3 = makeID({
  version: `3`,
  name: `segcryptor! test`,
  namespace: `2762ad30-9dc8-5cf2-8042-8cd411d450fc`,
});
console.log(`makeID : exampleMakeIDv3: ${exampleMakeIDv3}\r\n`);

/******************************************************************************/

// test generating a blend of UUIDv1 and UUIDv3
const exampleMakeIDv13 = makeID({
  version: `1.3`,
  name: `segcryptor! test`,
  namespace: `https://github.com/s3govesus/segcryptor`,
});
console.log(`makeID : exampleMakeIDv13: ${exampleMakeIDv13}\r\n`);

/******************************************************************************/

// test generating a blend of UUIDv3 and UUIDv1
const exampleMakeIDv31 = makeID({
  version: `3.1`,
  name: `segcryptor! test`,
  namespace: `https://github.com/s3govesus/segcryptor`,
});
console.log(`makeID : exampleMakeIDv31: ${exampleMakeIDv31}\r\n`);

/******************************************************************************/

// test generating a UUIDv4
const exampleMakeIDv4 = makeID({
  version: `4`,
});
console.log(`makeID : exampleMakeIDv4: ${exampleMakeIDv4}\r\n`);

/******************************************************************************/

// test generating a blend of UUIDv1 and UUIDv4
const exampleMakeIDv14 = makeID({
  version: `1.4`,
});
console.log(`makeID : exampleMakeIDv14: ${exampleMakeIDv14}\r\n`);

/******************************************************************************/

// test generating a blend of UUIDv4 and UUIDv1
const exampleMakeIDv41 = makeID({
  version: `4.1`,
});
console.log(`makeID : exampleMakeIDv41: ${exampleMakeIDv41}\r\n`);

/******************************************************************************/

// test generating a UUIDv5
const exampleMakeIDv5 = makeID({
  version: `5`,
  name: `segcryptor! test`,
  namespace: `https://github.com/s3govesus/segcryptor`,
});
console.log(`makeID : exampleMakeIDv5: ${exampleMakeIDv5}\r\n`);

/******************************************************************************/

// test generating a blend of UUIDv1 and UUIDv5
const exampleMakeIDv15 = makeID({
  version: `1.5`,
  name: `segcryptor! test`,
  namespace: `https://github.com/s3govesus/segcryptor`,
});
console.log(`makeID : exampleMakeIDv15: ${exampleMakeIDv15}\r\n`);

/******************************************************************************/

// test generating a blend of UUIDv5 and UUIDv1
const exampleMakeIDv51 = makeID({
  version: `5.1`,
  name: `segcryptor! test`,
  namespace: `https://github.com/s3govesus/segcryptor`,
});
console.log(`makeID : exampleMakeIDv51: ${exampleMakeIDv51}\r\n`);

/******************************************************************************/

// test generating a 256-bit abomination
const exampleMakeIDv0 = makeID({
  version: `0`,
  seed: `lalalalalalalalalala`,
});
console.log(`makeID : exampleMakeIDv0: ${exampleMakeIDv0}\r\n`);

/******************************************************************************/

// test generating a 256-bit abomination v01
const exampleMakeIDv01 = makeID({
  version: `0.1`,
  seed: `oh no, this is horrible!`,
});
console.log(`makeID : exampleMakeIDv01: ${exampleMakeIDv01}\r\n`);

/******************************************************************************/

const exampleMakeIDv10 = makeID({
  version: `1.0`,
  seed: `oh boy, here we go again!`,
});
console.log(`makeID : exampleMakeIDv10: ${exampleMakeIDv10}\r\n`);

/******************************************************************************/

const exampleMakeIDv03 = makeID({
  version: `0.3`,
  seed: `great...`,
  name: `this is an example name, i guess`,
  namespace: `https://www.google.com`,
});
console.log(`makeID : exampleMakeIDv03: ${exampleMakeIDv03}\r\n`);

/******************************************************************************/

const exampleMakeIDv30 = makeID({
  version: `3.0`,
  seed: `this is some seed!`,
  name: `names are dumb`,
  namespace: `https://namespacesaredumber.com`,
});
console.log(`makeID : exampleMakeIDv30: ${exampleMakeIDv30}\r\n`);

/******************************************************************************/

const exampleMakeIDv04 = makeID({
  version: `0.4`,
  seed: `what is a seed?`,
});
console.log(`makeID : exampleMakeIDv04: ${exampleMakeIDv04}\r\n`);

/******************************************************************************/

const exampleMakeIDv40 = makeID({
  version: `4.0`,
  seed: `broccoli smells a lot like farts - it's my favorite vegetable!`,
});
console.log(`makeID : exampleMakeIDv40: ${exampleMakeIDv40}\r\n`);

/******************************************************************************/

const exampleMakeIDv05 = makeID({
  version: `0.5`,
  seed: `i hate you, kind of...`,
  name: `it ain't what they call you, it's what you answer to`,
  namespace: `https://wcfields.com`,
});
console.log(`makeID : exampleMakeIDv05: ${exampleMakeIDv05}\r\n`);

/******************************************************************************/

const exampleMakeIDv50 = makeID({
  version: `5.0`,
  seed: `a seed neither fears light nor darkness, but uses both to grow`,
  name: `Matshona Dhliwayo`,
  namespace: `https://matshonadhliwayo.net`,
});
console.log(`makeID : exampleMakeIDv50: ${exampleMakeIDv50}\r\n`);

/******************************************************************************/

// test creating a random hexadecimal string
const exampleMakeHash = makeHash({
  size: 8,
  isSecure: true,
});
console.log(`makeHash : exampleMakeHash: ${exampleMakeHash}\r\n`);

/******************************************************************************/

// test generating a random alpha-numeric string
const exampleMakeKey = makeKey({
  size: 16,
  isComplex: true,
});
console.log(`makeKey : exampleMakeKey: ${exampleMakeKey}\r\n`);

/******************************************************************************/

// test combining a salt hash with a password hash and then encrypting the result
// the password should be hashed at least once, using the SHA-512 algorithm, which can be done by calling 'hashString('the password string', numberOfTimesToHash)'
// the salt should also be a 128-character hexadecimal string, like you would get after running SHA-512, or you could generate a random salt by using 'makeHash({size:64, isSecure:false})'
const passwordHash = `43b71277bc68e66a5f183da93211c0e6796a0ece455e926556c01de44a9131eba4d10dc8869e20a5d2631231f48b43e494a29182997ca7702f162a6f78349e0e`;
const salt = `9f3313fa38e30bd23db2445f5e57202f73dcc07d13c218b9e62cbeadd6da633f0e81defd6f8d013a7e556ac967cd2b5da0f876908ccb2b0e57fc8df41b948685`;
const exampleEncryptPassword = encryptPassword(passwordHash, saltHash, {
  security: 9, // this should run the SHA-512 algorithm 512 times to help prevent dictionary attacks
});
console.log(`passwordHash: ${passwordHash}`);
console.log(`salt: ${salt}`);
console.log(`encryptPassword : exampleEncryptPassword: ${exampleEncryptPassword}\r\n`);

/******************************************************************************/

// test applying salt to a password hash
const saltedHash = saltHash(passwordHash, salt);
console.log(`saltHash : saltedHash: ${saltedHash}\r\n`);

/******************************************************************************/

// test applying the SHA-512 algorithm to a string
const exampleHashValue512 = hashValue(`this is an example`, { count: 5, type: `sha512` });
console.log(`hashValue : exampleHashValue512: ${exampleHashValue512}\r\n`);

const exampleHashValue256 = hashValue(`this is an example`, { count: 12, type: `sha256` });
console.log(`hashValue : exampleHashValue256: ${exampleHashValue256}\r\n`);

const exampleHashValue512v2 = webHashValue(`this is an example`, { count: 7, type: `sha512` });
console.log(`webHashValue : exampleHashValue512v2: ${exampleHashValue512v2}\r\n`);

/******************************************************************************/

const exampleToBoolean1 = toBoolean(`true`);
console.log(`toBoolean : exampleToBoolean1 : ${exampleToBoolean1}`);

const exampleToBoolean2 = toBoolean(1);
console.log(`toBoolean : exampleToBoolean2 : ${exampleToBoolean2}`);

const exampleToBoolean3 = toBoolean(`off`);
console.log(`toBoolean : exampleToBoolean3 : ${exampleToBoolean3}`);

/******************************************************************************/

// test the encrypt() and decrypt() functions
function testEncryptDecrypt() {
  console.log(`Testing the encrypt() and decrypt() functions...`);
  console.log();

  let testStr = `This is a test.`;
  console.log(`testStr : ${testStr}`);

  let key = makeHash({
    size: 32,
  });
  console.log(`key : ${key}`);

  let iv = makeHash({
    size: 16,
  });
  console.log(`iv : ${iv}`);

  const encryptedStr = encrypt(testStr, key, iv);
  console.log(`encryptedStr : ${encryptedStr}`);
  let decryptedStr = decrypt(encryptedStr, key, iv);
  console.log(`decryptedStr : ${decryptedStr}`);
  console.log();
  console.log(`Finished testing the encrypt() and decrypt() functions!`);
  console.log(``);
  console.log(`/******************************************************************************/`);
  console.log(``);
}
testEncryptDecrypt();

/******************************************************************************/

// test the latinToHex function
function testLatinToHex() {
  console.log(`Testing the latinToHex() function...`);
  console.log();
  let testStr = `this is a test`;
  let hex = latinToHex(testStr);
  console.log(`this is a test : ${hex}`);
  console.log();
  console.log(`Finished testing the latinToHex() function!`);
  console.log();
  console.log(`/******************************************************************************/`);
  console.log();
}
testLatinToHex();

/******************************************************************************/

// test the hexToLatin function
function testHexToLatin() {
  console.log(`Testing the hexToLatin() function...`);
  console.log();
  let testStr = `7468697320697320612074657374`;
  let latin = hexToLatin(testStr);
  console.log(`7468697320697320612074657374 : ${latin}`);
  console.log();
  console.log(`Finished testing the hexToLatin() function!`);
  console.log();
  console.log(`/******************************************************************************/`);
  console.log();
}
testHexToLatin();


.eslintrc.js
module.exports = {
  env: {
    es2021: true,
  },
  extends: [
    `airbnb-base`,
  ],
  parserOptions: {
    ecmaVersion: 12,
    sourceType: `module`,
  },
  rules: {
    semi: [2, `always`],
    quotes: [2, `backtick`],
    'comma-dangle': [2, `always-multiline`],
    'prefer-const': 0,
    'max-len': 0,
    'no-underscore-dangle': 0,
    'spaced-comment': 0,
    'no-param-reassign': 0,
    'import/prefer-default-export': 0,
    'no-useless-return': 0,
    'no-shadow': 0,
    'no-unused-vars': 1,
    'guard-for-in': 0,
    'no-restricted-syntax': 0,
    'no-continue': 0,
    'no-case-declarations': 0,
    'no-unused-expressions': 1,
    'no-await-in-loop': 0,
    'no-loop-func': 0,
    'no-global-assign': 1,
    'global-require': 0,
    'prefer-destructuring': 0,
    'no-use-before-define': 0,
    'no-console': 0,
    'no-lonely-if': 0,
    'no-empty': 1,
  },
};


src/lib/id.js
const { v1: uuidv1 } = require(`uuid`);
// there is no uuidv2 function
const { v3: uuidv3 } = require(`uuid`);
const { v4: uuidv4 } = require(`uuid`);
const { v5: uuidv5 } = require(`uuid`);

const { hashValue, saltHash } = require(`./hash`);

const uuidRegex = /^([a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12})$/g;

/******************************************************************************/

// generates an ID usually made up of one or more of the UUID algorithms
//
// EXAMPLE OPTIONS
// const options = {
//   version: 4, // the version of UUID or segcryptor-specific ID to generate
//   name: `segcryptor default`, // UUIDv3 or UUIDv5 specific parameter
//   namespace: `https://github.com/s3govesus/segcryptor`, // UUIDv3 or UUIDv5 specific parameter
//   seed: `1582683512738` // version 0 / segcryptor-specific parameter of random data - uses Date.now().toString() by default
// };
function makeID(options) {
  let value = ``;

  // apply default settings if necessary
  try {
    if (options === undefined || typeof options !== `object`) {
      options = {
        version: 4,
        name: Date.now().toString(),
        namespace: `https://github.com/s3govesus/segcryptor`,
        seed: Date.now().toString(),
      };
    } else {
      if (options.version === undefined) {
        options.version = `4`;
      } else if (typeof options.version !== `string`) {
        throw new Error(`The value for the 'version' option in makeID() must be a string.`);
      }
      if (options.name === undefined) {
        options.name = Date.now().toString();
      } else {
        options.name = options.name.toString();
      }
      if (options.namespace === undefined) {
        options.namespace = `https://github.com/s3govesus/segcryptor`;
      } else {
        options.namespace = options.namespace.toString();
      }
      if (options.seed === undefined) {
        options.seed = Date.now().toString();
      } else {
        options.seed = options.seed.toString();
      }
    }
  } catch (ex) {
    // ? TODO
    throw new Error(ex.message);
  }

  // decide what kind of ID to generate based on the options parameter and do it
  switch (options.version) {
    case `1`:
      value = makev1();
      break;
    case `3`:
      value = makev3(options);
      break;
    case `1.3`:
      value = makev13(options);
      break;
    case `3.1`:
      value = makev31(options);
      break;
    case `4`:
      value = makev4();
      break;
    case `1.4`:
      value = makev14();
      break;
    case `4.1`:
      value = makev41();
      break;
    case `5`:
      value = makev5(options);
      break;
    case `1.5`:
      value = makev15(options);
      break;
    case `5.1`:
      value = makev51(options);
      break;
    case `0`: // randomly generated 256-bit uuid of my own design, with optional seed parameter
      value = makev0(options);
      break;
    case `0.1`:
      value = makev01(options);
      break;
    case `1.0`:
      value = makev10(options);
      break;
    case `0.3`:
      value = makev03(options);
      break;
    case `3.0`:
      value = makev30(options);
      break;
    case `0.4`:
      value = makev04(options);
      break;
    case `4.0`:
      value = makev40(options);
      break;
    case `0.5`:
      value = makev05(options);
      break;
    case `5.0`:
      value = makev50(options);
      break;
    default: // v4
      value = makev4();
      break;
  }

  return value;
}

/******************************************************************************/

function makev1() {
  let value = ``;

  value = uuidv1();

  return value;
}

/******************************************************************************/

function makev3(options) {
  let value = ``;

  // if the namespace isn't already a valid UUID namespace, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv3(options.namespace, uuidv3.URL);
    } else {
      options.namespace = uuidv3(options.namespace, uuidv3.DNS);
    }
  }
  // and then generate the actual result value UUID using that namespace + the name
  value = uuidv3(options.name, options.namespace);

  return value;
}

/******************************************************************************/

function makev13(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = uuidv1();
  // if the namespace is already a UUID
  // if the namespace isn't already a valid UUID, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv3(options.namespace, uuidv3.URL);
    } else {
      options.namespace = uuidv3(options.namespace, uuidv3.DNS);
    }
  }
  // and then generate the actual result value UUID using that namespace + the name
  part2 = uuidv3(options.name, options.namespace);
  part1 = part1.substring(0, 18);
  part2 = part2.substring(18, part2.length);
  value = part1 + part2;

  return value;
}

/******************************************************************************/

function makev31(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  // if the namespace isn't already a valid UUID, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv3(options.namespace, uuidv3.URL);
    } else {
      options.namespace = uuidv3(options.namespace, uuidv3.DNS);
    }
  }
  part1 = uuidv3(options.name, options.namespace);
  part2 = uuidv1();
  part1 = part1.substring(0, 18);
  part2 = part2.substring(18, part2.length);
  value = part1 + part2;

  return value;
}

/******************************************************************************/

function makev4() {
  let value = ``;

  value = uuidv4();

  return value;
}

/******************************************************************************/

function makev14() {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = uuidv1();
  part2 = uuidv4();
  part1 = part1.substring(0, 18);
  part2 = part2.substring(18, part2.length);
  value = part1 + part2;

  return value;
}

/******************************************************************************/

function makev41() {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = uuidv4();
  part2 = uuidv1();
  part1 = part1.substring(0, 18);
  part2 = part2.substring(18, part2.length);
  value = part1 + part2;

  return value;
}

/******************************************************************************/

function makev5(options) {
  let value = ``;

  // if the namespace isn't already a valid UUID, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv5(options.namespace, uuidv5.URL);
    } else {
      options.namespace = uuidv5(options.namespace, uuidv5.DNS);
    }
  }
  // and then generate the actual result value UUID using that namespace + the name
  value = uuidv5(options.name, options.namespace);

  return value;
}

/******************************************************************************/

function makev15(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = uuidv1();
  // if the namespace isn't already a valid UUID, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv5(options.namespace, uuidv5.URL);
    } else {
      options.namespace = uuidv5(options.namespace, uuidv5.DNS);
    }
  }
  // and then generate the actual result value UUID using that namespace + the name
  part2 = uuidv5(options.name, options.namespace);
  part1 = part1.substring(0, 18);
  part2 = part2.substring(18, part2.length);
  value = part1 + part2;

  return value;
}

/******************************************************************************/

function makev51(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  // if the namespace isn't already a valid UUID, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv5(options.namespace, uuidv5.URL);
    } else {
      options.namespace = uuidv5(options.namespace, uuidv5.DNS);
    }
  }
  part1 = uuidv5(options.name, options.namespace);
  part1 = part1.substring(0, 18);
  part2 = uuidv1();
  part2 = part2.substring(18, part2.length);
  value = part1 + part2;

  return value;
}

/******************************************************************************/

function makev0(options) {
  let value = ``;

  value = makeID256(options.seed);

  return value;
}

/******************************************************************************/

function makev01(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = makeID256(options.seed);
  part1 = part1.substring(0, 37);
  part2 = uuidv1();
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

function makev10(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = uuidv1();
  part2 = makeID256(options.seed);
  part2 = part2.substring(36, part2.length);
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

function makev03(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = makeID256(options.seed);
  part1 = part1.substring(0, 37);
  // if the namespace isn't already a valid UUID namespace, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv3(options.namespace, uuidv3.URL);
    } else {
      options.namespace = uuidv3(options.namespace, uuidv3.DNS);
    }
  }
  part2 = uuidv3(options.name, options.namespace);
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

function makev30(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  // if the namespace isn't already a valid UUID namespace, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv3(options.namespace, uuidv3.URL);
    } else {
      options.namespace = uuidv3(options.namespace, uuidv3.DNS);
    }
  }
  part1 = uuidv3(options.name, options.namespace);
  part2 = makeID256(options.seed);
  part2 = part2.substring(36, part2.length);
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

function makev04(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = makeID256(options.seed);
  part1 = part1.substring(0, 37);
  part2 = uuidv4();
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

function makev40(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = uuidv4();
  part2 = makeID256(options.seed);
  part2 = part2.substring(36, part2.length);
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

function makev05(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  part1 = makeID256(options.seed);
  part1 = part1.substring(0, 37);
  // if the namespace isn't already a valid UUID, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv3(options.namespace, uuidv3.URL);
    } else {
      options.namespace = uuidv3(options.namespace, uuidv3.DNS);
    }
  }
  part2 = uuidv5(options.name, options.namespace);
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

function makev50(options) {
  let value = ``;

  let part1 = ``;
  let part2 = ``;

  // if the namespace isn't already a valid UUID, make it one
  if (uuidRegex.test(options.namespace) === false) {
    if (options.namespace.indexOf(`://`) >= 0) {
      options.namespace = uuidv3(options.namespace, uuidv3.URL);
    } else {
      options.namespace = uuidv3(options.namespace, uuidv3.DNS);
    }
  }
  part1 = uuidv5(options.name, options.namespace);
  part2 = makeID256(options.seed);
  part2 = part2.substring(36, part2.length);
  value = `${part1}${part2}`;

  return value;
}

/******************************************************************************/

// generates a 256-bit ID instead of a 128-bit UUID, with an optional seed
// const resultEx = `d5820238-c6a7-11e9-aa8c-2a2ae2dbcce4-d5820238-c6a7-11e9-aa8c-2a2ae2dbcce4`
function makeID256(seed) {
  let result = ``;

  // generate a SHA-512 hash of the current unix timestamp
  let hash = Date.now().toString();
  hash = hashValue(hash, { count: 1 });

  if (seed === undefined) {
    seed = Date.now().toString();
  }
  seed = hashValue(seed, { count: 1 });

  hash = saltHash(hash, seed);
  hash = hashValue(hash, { count: 3 });

  // take the first 8 characters of the 128-character cryptographic hash
  result += hash.substring(0, 8);
  hash = hash.substring(8, hash.length);

  // generate 4 random hexidecimal characters (2x 16-bit values)
  let rand1 = ``;
  for (let i = 0; i < 2; i += 1) {
    const rando = Math.floor(Math.random() * 256);
    let randoString = rando.toString(16);
    if (randoString.length === 1) {
      randoString = `0${randoString}`;
    }
    rand1 += randoString;
  }
  result += `-${rand1}`;

  // take the next 4 characters from the cryptographic hash
  result += `-${hash.substring(0, 4)}`;
  hash = hash.substring(4, hash.length);

  // generate 4 random hexidecimal characters (2x 16-bit values)
  let rand2 = ``;
  for (let i = 0; i < 2; i += 1) {
    const rando = Math.floor(Math.random() * 256);
    let randoString = rando.toString(16);
    if (randoString.length === 1) {
      randoString = `0${randoString}`;
    }
    rand2 += randoString;
  }
  result += `-${rand2}`;

  // take the next 12 characters from the cryptographic hash
  result += `-${hash.substring(0, 12)}`;
  hash = hash.substring(12, hash.length);

  // generate 8 random hexidecimal characters (4x 16-bit values)
  let rand3 = ``;
  for (let i = 0; i < 4; i += 1) {
    const rando = Math.floor(Math.random() * 256);
    let randoString = rando.toString(16);
    if (randoString.length === 1) {
      randoString = `0${randoString}`;
    }
    rand3 += randoString;
  }
  result += `-${rand3}`;

  // take the next 4 characters from the cryptographic hash
  result += `-${hash.substring(0, 4)}`;
  hash = hash.substring(4, hash.length);

  // generate 4 random hexidecimal characters (2x 16-bit values)
  let rand4 = ``;
  for (let i = 0; i < 2; i += 1) {
    const rando = Math.floor(Math.random() * 256);
    let randoString = rando.toString(16);
    if (randoString.length === 1) {
      randoString = `0${randoString}`;
    }
    rand4 += randoString;
  }
  result += `-${rand4}`;

  // take the next 4 characters from the cryptographic hash
  result += `-${hash.substring(0, 4)}`;
  hash = hash.substring(4, hash.length);

  // generate 12 random hexidecimal characters (6x 16-bit values)
  let rand5 = ``;
  for (let i = 0; i < 6; i += 1) {
    const rando = Math.floor(Math.random() * 256);
    let randoString = rando.toString(16);
    if (randoString.length === 1) {
      randoString = `0${randoString}`;
    }
    rand5 += randoString;
  }
  result += `-${rand5}`;

  return result;
}

module.exports = {
  makeID,
};
